File: scripts/ingest-fsq.ts
/**
 * ingest-fsq.ts — uredniški ingest Foursquare OS Places → data/fsq-places/*.jsonl
 *
 * TASK 61 (1.61.0): DESETI aktivni sloj — LOKALNA odprta množica (Apache-2.0
 * z atribucijo). Dvostopenjska cev (poslovna logika SAMO tu, v TS — isti
 * moduli kot adapter, ni dvojnega vira resnice):
 *   1. scripts/ingest-fsq.py (python3 + duckdb httpfs) — „dumb pipe“:
 *      odkrije datoteke posnetka, ki se sekajo z regijo (parquet metapodatki
 *      — range-pushdown), izlušči vrstice v union bbox-u v vmesni JSONL (/tmp).
 *   2. TA skripta — filtri + preslikava v FsqPlace JSONL obliko:
 *      - regija: SUPPORTED_COUNTRY_BBOXES (SI+HR+ME+AL, kanonski približki);
 *      - obseg: fsqPlaceInScope (potovalno-relevantne kategorije — editorial);
 *      - zaprti kraji (date_closed) izločeni;
 *      - naslov sestavljen v formatted_address; kategorije → [{label}] (≤4);
 *      - izhod PO DRŽAVAH: data/fsq-places/{si,hr,me,al}.jsonl (git baseline
 *        — diff pokaže spremembe, uredniška kontrola, KT vzorec).
 *
 * Zagon:
 *   bun run fsq:ingest                       (privzeti posnetek 2025-02-06)
 *   bun run fsq:ingest -- --snapshot=2025-02-06 --keep-raw=/tmp/fsq-raw.jsonl
 *
 * Etika/licenca: Apache-2.0 z atribucijo (izpisuje mapper na vsakem
 * produktu); NE komercialni vir (info_only) — brez cen/razpoložljivosti.
 */

import { spawnSync } from "node:child_process";
import { createReadStream, existsSync, rmSync, writeFileSync } from "node:fs";
import { mkdir, rename } from "node:fs/promises";
import * as readline from "node:readline";
import path from "node:path";

import {
  SUPPORTED_COUNTRY_BBOXES,
  supportedCountryOf,
  fsqPlaceInScope,
  type SupportedCountryCode,
} from "../src/lib/supply/providers/fsq/dataset";
import { isFsqPlace, type FsqPlace } from "../src/lib/supply/providers/fsq/types";

const DEFAULT_SNAPSHOT = "2025-02-06";
const HERE = path.dirname(new URL(import.meta.url).pathname);
const PROJECT = path.resolve(HERE, "..");
const OUT_DIR = path.join(PROJECT, "data", "fsq-places");

/** Vmesna (surova) vrstica ekstraktorja — oblika distribucije fused.io. */
interface RawRow {
  fsq_id?: string;
  name?: string;
  latitude?: number;
  longitude?: number;
  address?: string | null;
  locality?: string | null;
  region?: string | null;
  postcode?: string | null;
  country?: string | null;
  tel?: string | null;
  website?: string | null;
  email?: string | null;
  date_refreshed?: string | null;
  date_closed?: string | null;
  categories?: string[] | null;
}

const str = (v: unknown): string | undefined =>
  typeof v === "string" && v.trim().length > 0 ? v.trim() : undefined;

/** Dodelitev države: PRIMARNO lastno polje `country` vira (ISO koda —
 *  avtoriteta vira, pravilno za Zagreb/Kotor/Tirano kljub prekrivanju
 *  pravokotnikov); NADOMESTNO kanonski bbox, kadar vir kode nima. */
function countryOfRaw(
  r: RawRow,
  lat: number,
  lng: number
): SupportedCountryCode | null {
  const raw = str(r.country);
  if (raw != null) {
    const cc = raw.toUpperCase();
    if (cc === "SI" || cc === "HR" || cc === "ME" || cc === "AL") {
      return cc as SupportedCountryCode;
    }
    return null; // vir prijavlja TUJO državo — izven regije (bbox NE povozí)
  }
  // vir NE prijavlja države (~11 od 1,17 M) → kanonska bbox rezerva.
  return supportedCountryOf(lat, lng);
}

/** Sestavi formatted_address IZKLJUČNO iz delov vira (NE izmišljujemo). */
function formattedAddress(r: RawRow): string | undefined {
  const street = str(r.address);
  const locality = str(r.locality);
  const country = str(r.country);
  const parts = [street, locality, country].filter((p): p is string => p != null);
  return parts.length > 0 ? parts.join(", ").slice(0, 300) : undefined;
}

/** Preslikava surove vrstice v FsqPlace (polja, ki jih mapper dejansko bere). */
function toPlace(r: RawRow): FsqPlace | null {
  const fsqId = str(r.fsq_id);
  const name = str(r.name);
  if (
    fsqId == null ||
    name == null ||
    typeof r.latitude !== "number" ||
    !Number.isFinite(r.latitude) ||
    typeof r.longitude !== "number" ||
    !Number.isFinite(r.longitude)
  ) {
    return null;
  }
  const cats = Array.isArray(r.categories)
    ? r.categories
        .filter((c): c is string => typeof c === "string" && c.trim().length > 0)
        .slice(0, 4)
        .map((label) => ({ label: label.slice(0, 160) }))
    : [];
  const place: FsqPlace = {
    fsq_id: fsqId,
    name,
    latitude: r.latitude,
    longitude: r.longitude,
    ...(cats.length > 0 ? { categories: cats } : {}),
    ...(formattedAddress(r) ? { address: { formatted_address: formattedAddress(r)! } } : {}),
    ...(str(r.tel) ? { tel: str(r.tel)!.slice(0, 60) } : {}),
    ...(str(r.website) ? { website: str(r.website)!.slice(0, 500) } : {}),
    ...(str(r.date_refreshed) ? { date_refreshed: str(r.date_refreshed) } : {}),
  };
  return isFsqPlace(place) ? place : null;
}

async function main() {
  const argSnapshot = process.argv.find((a) => a.startsWith("--snapshot="));
  const snapshot = argSnapshot ? argSnapshot.slice("--snapshot=".length) : DEFAULT_SNAPSHOT;
  const argKeep = process.argv.find((a) => a.startsWith("--keep-raw="));
  const argFromRaw = process.argv.find((a) => a.startsWith("--from-raw="));
  const rawPath =
    argFromRaw?.slice("--from-raw=".length) ??
    argKeep?.slice("--keep-raw=".length) ??
    path.join("/tmp", `fsq-raw-${Date.now()}.jsonl`);

  // Union bbox IZ kanonskih državnih bbox-ov (edini vir: TS modul).
  const boxes = Object.values(SUPPORTED_COUNTRY_BBOXES);
  const latMin = Math.min(...boxes.map((b) => b.latMin));
  const latMax = Math.max(...boxes.map((b) => b.latMax));
  const lngMin = Math.min(...boxes.map((b) => b.lngMin));
  const lngMax = Math.max(...boxes.map((b) => b.lngMax));
  console.log(
    `[fsq] regija: ${Object.keys(SUPPORTED_COUNTRY_BBOXES).join("+")} ` +
      `(lat ${latMin}–${latMax}, lng ${lngMin}–${lngMax}), posnetek ${snapshot}`
  );

  // 1) EKSTRAKCIJA (python3 + duckdb; dumb pipe) — razen če že imamo surovo
  //    množico (--from-raw: nadaljevanje/debug brez ponovnega prenosa).
  if (argFromRaw) {
    if (!existsSync(rawPath)) {
      console.error(`[fsq] --from-raw=${rawPath} NE OBSTOJA.`);
      process.exit(1);
    }
    console.log(`[fsq] 1/2 PRESKOČENA ekstrakcija (--from-raw=${rawPath}) …`);
  } else {
    console.log("[fsq] 1/2 ekstrakcija (python3 scripts/ingest-fsq.py — DuckDB httpfs pushdown) …");
    const py = spawnSync(
      "python3",
      [
        path.join(HERE, "ingest-fsq.py"),
        "--bbox",
        `${latMin},${latMax},${lngMin},${lngMax}`,
        "--snapshot",
        snapshot,
        "--out",
        rawPath,
      ],
      { stdio: "inherit", cwd: PROJECT }
    );
    if (py.status !== 0 || !existsSync(rawPath)) {
      console.error(`[fsq] NEUSPEH ekstrakcije (status ${py.status}) — nič NI nameščeno.`);
      process.exit(1);
    }
  }

  // 2) FILTRI + ZAPIS PO DRŽAVAH (TS poslovna logika).
  console.log("[fsq] 2/2 filtri (države + obseg + zaprti) in zapis po državah …");
  const byCountry = new Map<string, string[]>(); // vrstice JSONL
  let raw = 0;
  let skippedInvalid = 0;
  let skippedOutOfRegion = 0;
  let skippedClosed = 0;
  let skippedOutOfScope = 0;

  const rl = readline.createInterface({
    input: createReadStream(rawPath, "utf-8"),
    crlfDelay: Infinity,
  });
  for await (const line of rl) {
    const t = line.trim();
    if (t.length === 0) continue;
    raw++;
    let parsed: unknown;
    try {
      parsed = JSON.parse(t);
    } catch {
      skippedInvalid++;
      continue;
    }
    const r = parsed as RawRow;
    // zaprti kraji (distribucija nosi date_closed namesto bool)
    if (str(r.date_closed) != null) {
      skippedClosed++;
      continue;
    }
    const place = toPlace(r);
    if (place == null) {
      skippedInvalid++;
      continue;
    }
    // DRŽAVA: lastna koda vira primarno, bbox rezerva (Zagreb → HR kljub
    // prekrivanju s SI pravokotnikom; Dunaj/Beograd/Trst → izven).
    const country = countryOfRaw(r, place.latitude, place.longitude);
    if (country == null) {
      skippedOutOfRegion++;
      continue;
    }
    if (!fsqPlaceInScope(place)) {
      skippedOutOfScope++;
      continue;
    }
    const lines = byCountry.get(country) ?? [];
    lines.push(JSON.stringify(place, undefined, 0));
    byCountry.set(country, lines);
  }

  // Atomarna namestitev: .tmp → rename (nalagalnik vidi celo datoteko).
  await mkdir(OUT_DIR, { recursive: true });
  const order = ["SI", "HR", "ME", "AL"] as const;
  let total = 0;
  let bytes = 0;
  for (const code of order) {
    const lines = byCountry.get(code) ?? [];
    const file = path.join(OUT_DIR, `${code.toLowerCase()}.jsonl`);
    const tmp = `${file}.tmp`;
    writeFileSync(tmp, lines.length > 0 ? lines.join("\n") + "\n" : "", "utf-8");
    await rename(tmp, file);
    total += lines.length;
    bytes += lines.reduce((a, l) => a + l.length + 1, 0);
    console.log(`[fsq]   ${path.relative(PROJECT, file)}: ${lines.length.toLocaleString()} krajev`);
  }
  if (!argKeep) rmSync(rawPath, { force: true });

  console.log(
    `[fsq] MNOŽICA NAMEŠČENA: ${total.toLocaleString()} krajev ` +
      `(${(bytes / 1e6).toFixed(1)} MB) iz ${raw.toLocaleString()} surovih — ` +
      `zavrnjenih: ${skippedOutOfScope.toLocaleString()} izven obsega (editorial), ` +
      `${skippedClosed.toLocaleString()} zaprtih, ${skippedOutOfRegion.toLocaleString()} izven regije, ` +
      `${skippedInvalid.toLocaleString()} neveljavnih.`
  );
}

main().catch((e) => {
  console.error("[fsq] NAPAKA:", e);
  process.exit(1);
});

File: scripts/ingest-fsq.py
#!/usr/bin/env python3
# ============================================================================
# ingest-fsq.py — EKSTRAKTOR Foursquare OS Places (parquet → vmesni JSONL)
# TASK 61 (1.61.0)
# ============================================================================
# NAMEN: „dumb pipe“ ekstrakcija iz PRIMARNE distribucije fused.io
# (source.coop, S3 BREZ prijave — Apache-2.0 z atribucijo) z DuckDB httpfs
# range-pushdown (prenese SAMO row-groupe, ki se sekajo z regijo — ne 15 GB).
# Vsa POSLOVNA LOGIKA (države, kategorije, obseg) živi v
# scripts/ingest-fsq.ts (TS, isti moduli kot adapter) — ta skripta NE
# odloča o vsebini, samo izlušči vrstice v regiji.
#
# Zahteve: python3 + `pip install duckdb` (uporabljeno 1.5.x).
#
# Uporaba (kliče jo scripts/ingest-fsq.ts — lahko tudi ročno):
#   python3 scripts/ingest-fsq.py \
#     --bbox 39.6,46.9,13.3,21.1 \
#     --snapshot 2025-02-06 \
#     --out /tmp/fsq-raw.jsonl
#
# Izhod: JSONL — ena vrstica na kraj (SUROVA oblika distribucije):
#   { fsq_id, name, latitude, longitude, address(street), locality, region,
#     postcode, country, tel, website, email, date_refreshed, date_closed,
#     categories: [label, …] }
# Izpisi na stdout: števci (datoteke/row-groupi/vrstice) za dnevnik.
#
# Etika/licenca: Apache-2.0 odprti podatki (z atribucijo — atribucijo
# izpisuje mapper: „© Foursquare / Open Places Apache-2.0“). Zaprti kraji
# se prenesejo S (izloča jih TS pot); kategorije so label-poti vira.
# ============================================================================
import argparse
import json
import re
import sys
import time

import duckdb

SOURCE_COOP_LIST = "https://data.source.coop/fused/fsq-os-places"  # ?list-type=2 (S3 prefix list)
SOURCE_COOP_FILE = "https://data.source.coop/fused"  # + /fsq-os-places/<snapshot>/places/*.parquet
# Zrcalo (ista shema, en sam 10 GB parquet — fallback, če source.coop pade):
# https://huggingface.co/datasets/do-me/foursquare_places_100M/resolve/main/foursquare_places.parquet


def list_snapshot_files(snapshot: str) -> list[str]:
    """Seznam parquet datotek posnetka (S3 list, brez prijave)."""
    import urllib.request

    url = f"{SOURCE_COOP_LIST}?list-type=2"
    # UA: gol python-urllib CDN zavrača (403) — vljuden identiteten UA.
    req = urllib.request.Request(
        url, headers={"User-Agent": "Mozilla/5.0 (compatible; dsa-fsq-ingest/1.0)"}
    )
    with urllib.request.urlopen(req, timeout=60) as r:
        xml = r.read().decode("utf-8")
    keys = re.findall(r"<Key>([^<]+)</Key>", xml)
    files = [
        k for k in keys
        if k.startswith(f"fsq-os-places/{snapshot}/places/") and k.endswith(".parquet")
    ]
    return [f"{SOURCE_COOP_FILE}/{k}" for k in sorted(files, key=natural_key)]


def natural_key(key: str) -> list:
    """Numerično sortiranje (0,1,2,…,10 — ne leksikografsko)."""
    return [int(t) if t.isdigit() else t for t in re.split(r"(\d+)", key)]


def region_files(con, urls: list[str], lat_min, lat_max, lng_min, lng_max):
    """Odkrij datoteke, katerih row-groupi se sekajo z regijo (samo nogavice)."""
    hits = []
    for i, u in enumerate(urls):
        try:
            lat = {
                r[0]: (r[1], r[2])
                for r in con.execute(
                    f"SELECT row_group_id, CAST(stats_min_value AS DOUBLE), "
                    f"CAST(stats_max_value AS DOUBLE) FROM parquet_metadata('{u}') "
                    "WHERE path_in_schema='latitude' AND stats_min_value IS NOT NULL"
                ).fetchall()
            }
            lng = {
                r[0]: (r[1], r[2])
                for r in con.execute(
                    f"SELECT row_group_id, CAST(stats_min_value AS DOUBLE), "
                    f"CAST(stats_max_value AS DOUBLE) FROM parquet_metadata('{u}') "
                    "WHERE path_in_schema='longitude' AND stats_min_value IS NOT NULL"
                ).fetchall()
            }
        except Exception as e:  # napaka ene datoteke NE podre skena
            print(f"  [meta] PRESKOČENA {u.rsplit('/', 1)[1]}: {str(e)[:100]}", file=sys.stderr)
            continue
        n = 0
        for rg, (la0, la1) in lat.items():
            g = lng.get(rg)
            if g is None:
                continue
            if la0 <= lat_max and la1 >= lat_min and g[0] <= lng_max and g[1] >= lng_min:
                n += 1
        if n > 0:
            hits.append((u, n))
        if (i + 1) % 20 == 0:
            print(f"  [meta] …{i + 1}/{len(urls)} datotek, zadetki: {len(hits)}", file=sys.stderr)
    return hits


def main() -> int:
    ap = argparse.ArgumentParser(description="FSQ OS Places extractor (parquet → raw JSONL)")
    ap.add_argument("--bbox", required=True, help="latMin,latMax,lngMin,lngMax (regija)")
    ap.add_argument("--snapshot", default="2025-02-06", help="posnetek source.coop (datum)")
    ap.add_argument("--out", required=True, help="izhodna JSONL pot (vmesna, /tmp)")
    ap.add_argument("--memory", default="1500MB", help="meja pomnilnika DuckDB")
    args = ap.parse_args()

    lat_min, lat_max, lng_min, lng_max = (float(x) for x in args.bbox.split(","))
    print(f"[fsq-py] regija bbox: lat {lat_min}–{lat_max}, lng {lng_min}–{lng_max}")

    con = duckdb.connect()
    con.execute("INSTALL httpfs; LOAD httpfs;")
    con.execute(f"SET threads=4; SET memory_limit='{args.memory}';")

    t0 = time.time()
    urls = list_snapshot_files(args.snapshot)
    if not urls:
        print(f"[fsq-py] NAPAKA: posnetka {args.snapshot} ni na source.coop", file=sys.stderr)
        return 2
    print(f"[fsq-py] posnetek {args.snapshot}: {len(urls)} datotek ({time.time() - t0:.0f}s)")

    t1 = time.time()
    hits = region_files(con, urls, lat_min, lat_max, lng_min, lng_max)
    if not hits:
        print("[fsq-py] NAPAKA: nobena datoteka ne seka regije", file=sys.stderr)
        return 3
    print(
        f"[fsq-py] regija seka {len(hits)}/{len(urls)} datotek "
        f"({sum(n for _, n in hits)} row-groupov, {time.time() - t1:.0f}s): "
        + ", ".join(u.rsplit('/', 1)[1] for u, _ in hits)
    )

    t2 = time.time()
    lst = ", ".join(f"'{u}'" for u, _ in hits)
    q = f"""
    SELECT fsq_place_id, name, latitude, longitude, address, locality, region,
           postcode, country, tel, website, email, date_refreshed, date_closed,
           fsq_category_labels
    FROM read_parquet([{lst}])
    WHERE latitude BETWEEN {lat_min} AND {lat_max}
      AND longitude BETWEEN {lng_min} AND {lng_max}
    """
    written = 0
    with open(args.out, "w", encoding="utf-8") as f:
        cur = con.execute(q)
        while True:
            rows = cur.fetchmany(50_000)
            if not rows:
                break
            for row in rows:
                cats = row[14] if row[14] else []
                rec = {
                    "fsq_id": row[0],
                    "name": row[1],
                    "latitude": row[2],
                    "longitude": row[3],
                    "address": row[4],
                    "locality": row[5],
                    "region": row[6],
                    "postcode": row[7],
                    "country": row[8],
                    "tel": row[9],
                    "website": row[10],
                    "email": row[11],
                    "date_refreshed": row[12],
                    "date_closed": row[13],
                    "categories": cats,
                }
                f.write(json.dumps(rec, ensure_ascii=False, separators=(",", ":")) + "\n")
                written += 1
        f.flush()
    print(
        f"[fsq-py] ekstrakcija: {written:,} vrstic → {args.out} ({time.time() - t2:.0f}s, "
        f"skupaj {time.time() - t0:.0f}s)"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
